Moonlight macOS/Input/Control.m

```diff
@@ -0,0 +1,188 @@
+            case 4: //UP
+                [_controllerSupport clearButtonFlag:_controller flags:UP_FLAG];
+                break;
+            case 5: //RIGHT
+                [_controllerSupport clearButtonFlag:_controller flags:RIGHT_FLAG];
+                break;
+            case 6: //DOWN
+                [_controllerSupport clearButtonFlag:_controller flags:DOWN_FLAG];
+                break;
+            case 7: //LEFT
+                [_controllerSupport clearButtonFlag:_controller flags:LEFT_FLAG];
+                break;
+            case 10: //LB
+                [_controllerSupport clearButtonFlag:_controller flags:LB_FLAG];
+                break;
+            case 11: //RB
+                [_controllerSupport clearButtonFlag:_controller flags:RB_FLAG];
+                break;
+            case 12: //Y
+                [_controllerSupport clearButtonFlag:_controller flags:Y_FLAG];
+                break;
+            case 13: //B
+                [_controllerSupport clearButtonFlag:_controller flags:B_FLAG];
+                break;
+            case 14: //A
+                [_controllerSupport clearButtonFlag:_controller flags:A_FLAG];
+                break;
+            case 15: //X
+                [_controllerSupport clearButtonFlag:_controller flags:X_FLAG];
+                break;
+                
+            default:
+                break;
+        }
+        [_controllerSupport updateFinished:_controller];
+    }
+}
+
+void onAxisMoved(struct Gamepad_device * device, unsigned int axisID, float value, float lastValue, double timestamp, void * context) {
+    if (verbose && /*(axisID <= 4) &&*/ fabsf(lastValue - value) > 0.01) {
+        switch (axisID) {
+            case 0: //y-Axis of Right Stick
+                printf("%u", axisID);
+                _controller.lastLeftStickX = value * 0X7FFE;
+                break;
+            case 1: //x-Axis of Left Stick
+                printf("%u", axisID);
+                _controller.lastLeftStickY = -value * 0X7FFE;
+                break;
+            case 2: //X-Axis of Right Stick
+                printf("%u", axisID);
+                _controller.lastRightStickX = value * 0X7FFE;
+                break;
+            case 3: //Y-Axis of Right Stick
+                printf("%u", axisID);
+                _controller.lastRightStickY = -value * 0X7FFE;
+                break;
+            case 14:
+                _controller.lastLeftTrigger = value * 0xFF;
+                break;
+            case 15:
+                _controller.lastRightTrigger = value * 0xFF;
+                break;
+            default:
+                break;
+        }
+        [_controllerSupport updateFinished:_controller];
+    }
+}
+
+void onDeviceAttached(struct Gamepad_device * device, void * context) {
+    if (verbose) {
+        printf("Device ID %u attached (vendor = 0x%X; product = 0x%X) with context %p\n", device->deviceID, device->vendorID, device->productID, context);
```

Comment:
You should create a `Controller` instance for each `struct Gamepad_device`, store it in an `NSMutableDictionary` and use that in the input event callbacks, rather than a single `Controller` object. This will give you proper multi-controller support. See iOS's ControllerSupport.m.

Reply:
This is a good idea, i can implement that, but have no way of testing it because i only have one controller. Or is there a virtual controller tool for testing?